### PR TITLE
Agent: Improve codeActions & diagnostics

### PR DESCRIPTION
**Pull request by RXminuS**
WIP:

- Consolidate protocol conversion methods into factories
- Ensure code-actions have more fine-grained `kind` types instead of matching on title
- Ensure codeAction/trigger handles non 'edit' use-cases

Note:
- Part of [ongoing changes](https://linear.app/sourcegraph/issue/CODY-3438/code-actions-bug-fixes-and-polish) on Jetbrains Client 
- Breaking protocol changes.
- Changed labels. Maybe needs to be separate PR, needs design approval / feedback

## Test plan
Do Not Merge Yet

### CODE REVIEW

**Comment by RXminuS** [comment]
* **#5314** <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/5314?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @RXminuS and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: vscode/src/jsonrpc/agent-protocol-factory/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Currently there are losts of different protocol constructors, conversion helpers, and manual interface allocations across `lib`, `agent`, `shim` and `vscode`. This has led to numerous bugs because a small protocol or implementation changes can have many side-effects that easily fly under the radar of the type-checker.
+
+Ultimately I'd like to move the entire protocol to a [TypeSpec](1) and generate these Protocol helpers/factories automatically. Also making sure the auto-generated code has convenient extension points to inject custom behavior. [Somewhat similar](2) to what we already (plan to) do with SCIP and the [language bindings in JetBrains.](3)
```
Comment (olafurpg) [inline vscode/src/jsonrpc/agent-protocol-factory/README.md:3]: FWIW, I think this is unlikely to happen. TypeSpec is pretty bare bones wrt code generation options (it generates OpenAPI and JSON SCHEMA) and OpenAPI generators are heavily optimized for simple API server/client bindings. Everything can be customized, but so can our SCIP tooling (and it doesn't have to be more complicated).

File: vscode/src/jsonrpc/agent-protocol-factory/position.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import assert from 'node:assert'
+import * as vscode from 'vscode'
+import type * as protocol from '../agent-protocol'
+
+export namespace Position {
+    export function from(position: vscode.Position): protocol.Position {
```
Comment (olafurpg) [inline vscode/src/jsonrpc/agent-protocol-factory/position.ts:6]: I believe this is redundant at the type level, but helpful for serialization. I think we should try to delay serialization-motivated conversions as much as possible.

File: vscode/src/jsonrpc/agent-protocol.ts [hunk 3]
```diff
@@ -1078,11 +1080,38 @@ export interface ProtocolCodeAction {
...
+     * - extract
+     * - inline
+     * - move
+     * - rewrite
+     * - source
+     * - organizeImports
+     * - fixAll
+     * - notebook
+     */
     kind?: string | undefined | null
```
Comment (olafurpg) [inline vscode/src/jsonrpc/agent-protocol.ts:1115]: Nit: im wondering if "featureID" or something like that is a better name. Fine to keep "kind", it's just an overloaded term and doesn't precisely describe what this string is.

**Review by olafurpg — CHANGES_REQUESTED** [review]
Overall LGTM for the direction. This is a nice cleanup, and makes sense to establish stronger conventions around the code structure here.

**Review by olafurpg — APPROVED** [review]

**Pull request closed without merging**